crates/matrix-sdk-crypto/src/olm/session.rs

```diff
@@ -156,11 +151,12 @@ impl Session {
                 recipient_device.ed25519_key().ok_or(EventError::MissingSigningKey)?;
 
             let payload = json!({
-                "sender": &self.user_id,
-                "sender_device": &self.device_id,
+                "sender": &self.our_device_keys.user_id,
+                "sender_device": &self.our_device_keys.device_id,
                 "keys": {
-                    "ed25519": self.our_identity_keys.ed25519.to_base64(),
+                    "ed25519": self.our_device_keys.ed25519_key().expect("Device doesn't have ed25519 key").to_base64(),
                 },
+                "device_keys": self.our_device_keys,
```

Comment:
Doesn't this property need a different name:

>Until this MSC is accepted, the new property should be named org.matrix.msc4147.device_keys.

We can do this in a follow up, so we don't risk more merge conflicts.

Reply:
https://github.com/matrix-org/matrix-rust-sdk/pull/3618